ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -3101,10 +3102,25 @@ private void constructOneLBLocationMap(FileStatus fSta,
       // For acid table, add the acid_write event with file list at the time of load itself. But
       // it should be done after partition is created.
 
+      List<WriteNotificationLogRequest> requestList = new ArrayList<>();
       for (Entry<Path, PartitionDetails> entry : partitionDetailsMap.entrySet()) {
         PartitionDetails partitionDetails = entry.getValue();
         if (isTxnTable && partitionDetails.newFiles != null) {
-          addWriteNotificationLog(tbl, partitionDetails.fullSpec, partitionDetails.newFiles, writeId);
+          addWriteNotificationLog(tbl, partitionDetails.fullSpec, partitionDetails.newFiles,
+                  writeId, requestList);
+        }
+      }
+      if (requestList.size() != 0) {
+        WriteNotificationLogBatchRequest rqst = new WriteNotificationLogBatchRequest(tbl.getCatName(), tbl.getDbName(),
+                tbl.getTableName(), requestList);
+        try {
+          get(conf).getSynchronizedMSC().addWriteNotificationLogInBatch(rqst);
+        } catch (TException e) {
+          // For older HMS, if the batch API is not supported, fall back to older API.
+          LOG.info("addWriteNotificationLogInBatch failed with ", e);
```

Comment:
Testing once for call being idempotent in case of failure should be helpful as every exception, like say for example, SQL Exception, it might be treated as incompatibility in HS2 and HMS versions.

Reply:
I have changed the code to retry only if the error is not supported method. And checked the idempotent behaviour also.